[OSDOCS-8025]: Azure confidential computing via MAPI (#65764)
openshift/openshift-docs: modules/machineset-azure-confidential-vms.adoc

// Module included in the following assemblies:
//
// * machine_management/creating_machinesets/creating-machineset-azure.adoc
// * machine_management/control_plane_machine_management/cpmso-using.adoc

ifeval::["{context}" == "cpmso-using"]
:cpmso:
endif::[]

:_content-type: PROCEDURE
[id="machineset-azure-confidential-vms_{context}"]
= Configuring Azure confidential virtual machines by using machine sets

{product-title} {product-version} supports Azure confidential virtual machines (VMs).

[NOTE]
====
Confidential VMs are currently not supported on 64-bit ARM architectures.
====

> By editing the machine set YAML file, you can configure the confidential VMs options that a machine set uses for machines that it deploys. These machines can use UEFI security features such as Secure Boot or a dedicated virtual Trusted Platform Module (vTPM) instance.

Reviewer: ```suggestion
By editing the machine set YAML file, you can configure the confidential VMs' options that a machine set uses for machines that it deploys. For example, you can configure these machines to use UEFI security features such as Secure Boot or a dedicated virtual Trusted Platform Module (vTPM) instance.
```
It's not clear to me what the logical connection is between these two sentences. Is the second sentence an example of options that can be set? Or is it just additional info? 

Consider rephrasing to tighten up the logical flow. My suggested edit might not be accurate, but maybe gives you a sense of what I'm talking about.

ifdef::cpmso[]

> [IMPORTANT]

Reviewer: Maybe consider changing to a "Warning" admonition, which, [according to the SSG](https://redhat-documentation.github.io/supplementary-style-guide/#admonitions), contains "Information about potential system damage, data loss, or a support-related issue if the user disregards this admonition." An unstable cluster could lead to data loss, correct?

====
Not all instance types support confidential VMs. Do not change the instance type for a control plane machine set that is configured to use confidential VMs to a type that is incompatible. Using an incompatible instance type can cause your cluster to become unstable.
====
endif::cpmso[]

For more information about related features and functionality, see the Microsoft Azure documentation about link:https://learn.microsoft.com/en-us/azure/confidential-computing/confidential-vm-overview[Confidential virtual machines].

.Procedure

> . In a text editor, open the YAML file for an existing machine set or create a new one.

Reviewer: ```suggestion
. In a text editor, create a new YAML file or open the YAML file for an existing machine set.
```
- In the current sentence structure, there's a chance that "create a new one" could be taken to refer to "machine set" rather than "file" because "set" is the closest noun in the sentence.

Author: Thy would create a new machine set by creating a new YAML file, so that's ok I think :slightly_smiling_face:

. Edit the following section under the `providerSpec` field:
+
--
.Sample configuration
[source,yaml]
----
ifndef::cpmso[]
apiVersion: machine.openshift.io/v1beta1
kind: MachineSet
endif::cpmso[]
ifdef::cpmso[]
apiVersion: machine.openshift.io/v1
kind: ControlPlaneMachineSet
endif::cpmso[]
# ...
spec:
  template:
    spec:
      providerSpec:
        value:
          osDisk:
            # ...
            managedDisk:
              securityProfile: # <1>
                securityEncryptionType: VMGuestStateOnly # <2>
            # ...
          securityProfile: # <3>
            settings:
                securityType: ConfidentialVM # <4>
                confidentialVM:
                  uefiSettings: # <5>
                    secureBoot: Disabled # <6>
                    virtualizedTrustedPlatformModule: Enabled # <7>
          vmSize: Standard_DC16ads_v5 # <8>
# ...
----
<1> Specifies security profile settings for the managed disk when using a confidential VM.
<2> Specifies the type of encryption used for a confidential VM. The following values are valid:
+

> * `DiskWithVMGuestState` enables encryption of the manged disk and the Azure VM Guest State (VMGS) blob with a platform-managed key (PMK) or a customer-managed key (CMK). This setting requires the use of vTPM and secure boot.

Reviewer: ```suggestion
* Setting this value to `DiskWithVMGuestState` enables encryption of the managed disk and the Azure VM Guest State (VMGS) blob with a platform-managed key (PMK) or a customer-managed key (CMK). This setting requires the use of vTPM and secure boot.
```
- Consider adding phrasing that indicates that this is a possible value rather than a field.

> * `VMGuestStateOnly` enables encryption of the VMGS blob. This setting requires the use of vTPM.

Author: Is the PMK/CMK statement applicable to both options?

Reviewer 2: The PMK/CMK statement is only applicable to `DiskWithVMGuestState`.

> * `VMGuestStateOnly` enables encryption of the VMGS blob. This setting requires the use of vTPM.

Reviewer: ```suggestion
* Setting this value to `VMGuestStateOnly` enables encryption of the VMGS blob. This setting requires the use of vTPM.
```
- Consider listing this item first because it's the one included in the sample code.

> <3> Specifies security profile settings for the managed disk when using a confidential VM.

Author: Should `<3>` also be "for the managed disk" (like `<1>`)?

Reviewer 2: `<3>` refers to the `securityProfile` of the Virtual Machine, while `<1>` refers to the `securityProfile` of the managed disk.

> I understand that it could be confusing, but this is how the Azure API is defined and we decided not to diverge too much from it.

Author: Thanks :) confusion is part of why I asked - I think if we label them clearly it will be as good as it can be. Will update with that.

<4> Enables the use of confidential VMs. This value is required for all valid configurations.
<5> Specifies which UEFI security features to use. This section is required for all valid configurations.
<6> Disables UEFI Secure Boot.
<7> Enables the use of a vTPM.
<8> Specifies an instance type that supports confidential VMs.
--

.Verification

> * On the Azure portal, review the details for a machine deployed by the machine set and verify that the confidential VMs options match the values that you configured.

Reviewer: ```suggestion
* On the Azure portal, review the details for a machine deployed by the machine set and verify that the confidential VMs' options match the values that you configured.
```
Not sure whether this should be singular or plural possessive.
